mibidata/combine_tiffs.py

```diff
@@ -0,0 +1,204 @@
+            the output to MIBItracker, the run label set here must match the
+            label of the MIBItracker run. Defaults to the name of the run xml.
+        instrument: Optionally, the instrument ID.
+        tissue: Optionally, the name of tissue.
+        aperture: Optionally, the name of the aperture or imaging preset.
+        out: Optionally, a path to a location for saving the combined TIFF. If
+           not specified, defaults to 'combined.tiff' inside the input folder.
+        run_label: Optionally, a custom run label for the `run` property of the
+            image.
+    """
+    panel_df = panels.read_csv(panel_path)
+    panel_name, _ = os.path.splitext(os.path.basename(panel_path))
+    tiff_files = os.listdir(input_folder)
+
+    fovs, calibration = runs.parse_xml(run_path)
+    point_number = int(point[5:])
+    try:
+        fov = fovs[point_number]
+    except IndexError:
+        raise IndexError('Point{} not found in run xml.'.format(point_number))
+    if fov['date']:
+        run_date = datetime.datetime.strptime(
+            fov['date'], '%Y-%m-%dT%H:%M:%S').date()
+    else:
+        run_date = datetime.datetime.now().date()
+
+    image_data = []
+    for i in panel_df.index:
+        tiff_path = os.path.join(
+            input_folder, _match_target_filename(tiff_files,
+                                                 panel_df['Target'][i]))
+        data = _load_single_channel(tiff_path)
+        image_data.append(data)
+
+    image_data = np.stack(image_data, axis=2)
+
+    image = mi.MibiImage(image_data,
+                         list(zip(panel_df['Mass'], panel_df['Target'])))
+
+    image.size = int(size)
+    image.coordinates = (fov['coordinates'])
+    image.filename = fov['run']
+    image.run = run_label if run_label else fov['run']
+    image.version = tiff.SOFTWARE_VERSION
+    image.instrument = instrument
+    image.slide = slide
+    image.dwell = fov['dwell']
+    image.scans = fov['scans']
+    image.aperture = aperture
+    image.point_name = fov['point_name']
+    image.folder = fov['folder']
+    image.tissue = tissue
+    image.panel = panel_name
+    image.date = run_date
+    image.mass_offset = calibration['MassOffset']
+    image.mass_gain = calibration['MassGain']
+    image.time_resolution = calibration['TimeResolution']
+
+    if out is None:
+        out = os.path.join(input_folder, 'combined.tiff')
+
+    tiff.write(out, image, multichannel=True)
+
+
+if __name__ == '__main__':
+
+    description = ('Generates a single multiplexed MIBItiff from a folder '
+                   'containing individual single-channel TIFFs without MIBI '
+                   'metadata. Note that if the single-channel TIFFs already '
+                   'have the MIBItiff metadata, then a much simpler way to '
+                   'merge them is to use combine_tiffs.merge_mibitiffs '
+                   'instead.')
+
+    parser = argparse.ArgumentParser(description=description)
```

Comment:
If I only specify the required args and try to open with BioFormats in ImageJ, I get a NullPointerException (stack below). Adding in instrument, tissue, and aperture fixes this so one or all of these may need some default values if not specified.

(Fiji Is Just) ImageJ 2.0.0-rc-65/1.51w; Java 1.8.0_66 [64-bit]; Windows 10 10.0; 108MB of 12156MB (<1%)
 
java.lang.NullPointerException
at loci.formats.in.IonpathMIBITiffReader.formatMetadata(IonpathMIBITiffReader.java:233)
at loci.formats.in.IonpathMIBITiffReader.initMetadataStore(IonpathMIBITiffReader.java:212)
at loci.formats.in.BaseTiffReader.initMetadata(BaseTiffReader.java:98)
at loci.formats.in.BaseTiffReader.initFile(BaseTiffReader.java:584)
at loci.formats.FormatReader.setId(FormatReader.java:1397)
at loci.plugins.in.ImportProcess.initializeFile(ImportProcess.java:499)
at loci.plugins.in.ImportProcess.execute(ImportProcess.java:142)
at loci.plugins.in.Importer.showDialogs(Importer.java:140)
at loci.plugins.in.Importer.run(Importer.java:76)
at loci.plugins.LociImporter.run(LociImporter.java:78)
at ij.IJ.runUserPlugIn(IJ.java:221)
at ij.IJ.runPlugIn(IJ.java:185)
at ij.Executer.runCommand(Executer.java:137)
at ij.Executer.run(Executer.java:66)
at java.lang.Thread.run(Thread.java:745)

Reply:
That is an excellent observation -- it looks like instrument is assumed to be present:
https://github.com/openmicroscopy/bioformats/blob/69499414f5502dbe24126f5770916ebe9d1086bc/components/formats-gpl/src/loci/formats/in/IonpathMIBITiffReader.java#L213

I have added a default value for that to the tiff writer so it's there not only when combining TIFFs but in general.

Reply:
Also, we should add a test using bioformats. Adding a ticket for that now.